### ISSUE TITLE
Error in gradient computation

### ISSUE THREAD

**Issue by pierrePalud** [pierrePalud/beetroots#2]
The code raises an error in an application with `N = 1` pixel, a neural network trained with `D_no_kappa = 2` physical parameters and `D_sampling = 3` (adding the scaling factor). The error is a dimension issue in the gradient computations obtained in the `NeuralNetworkApprox.compute_all()` method, an error that appears in the assert line 213:

```python
assert grad_log_f_Theta.shape == (
    N_pix,
    self.D_sampling,
    self.L,
), f"{grad_log_f_Theta.shape} is not ({N_pix}, {self.D_sampling}, {self.L})".
``` 

The error being `(1, 2, 101) is not (1, 3, 101)`.

The problem is that the computation of the gradient is:

```python
_Theta = torch.from_numpy(Theta_combined[:, 1:]) 

# ...

grad_log_f_Theta = (
    self.jacobian_network(_Theta)
    .detach()
    .numpy()  # .to(self.network.device)
).transpose(
    (0, 2, 1)
)  # (N, D, L)

assert grad_log_f_Theta.shape == (
    N_pix,
    self.D_sampling,
    self.L,
), f"{grad_log_f_Theta.shape} is not ({N_pix}, {self.D_sampling}, {self.L})".

# ... 

grad_log_f_Theta = grad_log_f_Theta[:, :, :] * self.LOGE_10
```

Here, the grad is evaluated for the full `theta` vector (that is, including those that are fixed / not inferred) and the scaling factor is currently not added. To correct the gradient:
- [ ] separate the true gradient to be returned, `grad_log_f_Theta`, from the gradient from `pytorch` (name it, eg, `grad_log_f_Theta_torch`)
- [ ] initialize `grad_log_f_Theta` with the correct shape
- [ ] add the gradient wrt the scaling factor (=1 if sampled in log scale) at `grad_log_f_Theta[:, 0, :]` 
- [ ]  select the indices of the subset of inferred physical parameters with `grad_log_f_Theta[:, 1:, :] = grad_log_f_Theta_torch[:, self.list_indices_to_sample_for_nn, :]`
- [ ] move the assert for the true gradient `grad_log_f_Theta` to after its definition
- [ ] Do the same for the 2nd order derivative `hess_full_log_f_Theta`.

**Comment by pierrePalud** [comment]
I regrouped the computation of this intermediate torch gradient and its transformation to the gradient to be used in optimization. The code now ensures the shape of the final gradient only.
